### DISCUSSION TITLE
about EVP_PKEY_CTX_new ...

### DISCUSSION THREAD

**Discussion by CandyMi** [openssl/openssl#22456] [Q&A]
version: OpenSSL 3.0.9 30 May 2023

```c++
  int ret = 0; EVP_PKEY *p1 = nullptr; EVP_PKEY *p2 = nullptr;

  /* paramgen*/
  {
    std::shared_ptr<EVP_PKEY_CTX> pctx(EVP_PKEY_CTX_new_id(EVP_PKEY_DSA, nullptr), EVP_PKEY_CTX_free);
    EVP_PKEY_paramgen_init(pctx.get());
    EVP_PKEY_CTX_set_dsa_paramgen_bits(pctx.get(), bits);
    // EVP_PKEY_CTX_set_dsa_paramgen_md(pctx.get(), EVP_sha1());
    ret = EVP_PKEY_paramgen(pctx.get(), &p1);
    if (ret != 1)
    {
      return handle_error("Failed to generate DSA parameter");
    }
  }

  /* keygen */
  {
    std::shared_ptr<EVP_PKEY_CTX> pctx(EVP_PKEY_CTX_new(p1, nullptr), EVP_PKEY_CTX_free);
    EVP_PKEY_keygen_init(pctx.get());
    ret = EVP_PKEY_keygen(pctx.get(), &p2);
    if (ret != 1)
    {
      return handle_error("Failed to generate DSA parameter");
    }
  }

  /* Do something with p2 ... */
```

My question is:

   * After calling `EVP_PKEY_CTX_new(p1, nullptr)`, does the `p1` pointer need to be released manually?

   * Why does using `EVP_PKEY_CTX_set_dsa_paramgen_md(pctx.get(), EVP_sha1())` cause `EVP_PKEY_paramgen` to fail?

Please forgive my presumptuous question, I haven't found the answer elsewhere yet.  : )

**Comment by t8m** [comment]
> My question is:
> 
>     * After calling `EVP_PKEY_CTX_new(p1, nullptr)`, does the `p1` pointer need to be released manually?

Yes

>     * Why does using `EVP_PKEY_CTX_set_dsa_paramgen_md(pctx.get(), EVP_sha1())` cause `EVP_PKEY_paramgen` to fail?

Most probably because you're generating DSA parameters with more than 2047 bits. SHA224 or SHA256 needs to be used otherwise.

**Reply by CandyMi** [reply]
I'm sorry for not replying to you for a long time!

I've tried making some modifications, but the error still exists:

```c++
EVP_PKEY_CTX_set_dsa_paramgen_bits(pctx.get(), 2048);
EVP_PKEY_CTX_set_dsa_paramgen_md(pctx.get(), EVP_sha256());
/* or */
EVP_PKEY_CTX_set_dsa_paramgen_bits(pctx.get(), 1024);
EVP_PKEY_CTX_set_dsa_paramgen_md(pctx.get(), EVP_sha1());
```
and error in `ret = EVP_PKEY_keygen(pctx.get(), &p2);`.

some print got `ret = 0`.

**Comment by slontis** [comment]
I would recommend running the sample in demos/signature/EVP_DSA_Signature. Refer to the Makefile for how to build and run
